packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,115 @@
+import type {
+  AsIntrinsicElement,
+  SlotComponent,
+  SlotComponentMetadata,
+  SlotRenderFunction,
+  SlotShorthandValue,
+  UnknownSlotProps,
+} from './types';
+import { SLOT_COMPONENT_METADATA_SYMBOL } from './constants';
+import * as React from 'react';
+import { isSlot } from './isSlot';
+
+export type SlotOptions<Props extends UnknownSlotProps> = {
+  elementType:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  defaultProps?: Partial<Props>;
+};
+
+/**
+ * Creates a slot from a slot shorthand or properties (`props.SLOT_NAME` or `props` itself)
+ * @param value - the value of the slot, it can be a slot shorthand, a slot component or a slot properties
+ * @param options - optional values you can pass to alter the signature of a slot, those values are:
+ *
+ * * `elementType` - the base element type of a slot, defaults to `'div'`
+ * * `defaultProps` - similar to a React component declaration, you can provide a slot default properties to be merged with the shorthand/properties provided
+ * * `required` - a boolean that indicates if a slot will be rendered even if it's base value is `undefined`.
+ * By default if `props.SLOT_NAME` is `undefined` then `state.SLOT_NAME` becomes `undefined`
+ * and nothing will be rendered, but if `required = true` then `state.SLOT_NAME` becomes an object
+ * with the values provided by `options.defaultProps` (or `{}`). This is useful for cases such as providing a default content
+ * in case no shorthand is provided, like the case of the `expandIcon` slot for the `AccordionHeader`
+ *
+ * @example of a required nullable slot
+ * ```tsx
+ * // AccordionHeader.types.ts
+ * type AccordionHeaderSlots = {
+ *    expandIcon?: Slot<'span'>
+ * }
+ * // useAccordionHeader.ts
+ * const state = {
+ *  expandIcon: slot(expandIconShorthand, {
+ *    required: true,
+ *    elementType: 'span',
+ *    defaultProps: { children: <ChevronRight/>, 'aria-hidden': true}
+ *  })
+ * }
+ * // renderAccordionHeader
+ * state.expandIcon && <state.expandIcon/>
+ * ```
+ */
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined,
+  options: { required: true } & SlotOptions<Props>,
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined | null,
+  options: { required?: boolean } & SlotOptions<Props>,
+): SlotComponent<Props> | undefined;
+export function slot<Props extends UnknownSlotProps>(
+  value: SlotComponent<Props>,
+  options?: { required: true } & Partial<SlotOptions<Props>>,
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined | null,
+  options: { required?: boolean } & Partial<SlotOptions<Props>> = {},
+): SlotComponent<Props> | undefined {
```

Comment:
When the `required` parameter is a conditional this approach becomes quite cumbersome, take the case of `DialogTitle.action` slot  for example, `required` is based on the presence of the `modalType` property:

```ts
slot(action, {
  required: modalType === 'non-modal',
  defaultProps: {
    // bunch of props...
  }
})
```

In this case with this `slot.*` approach it would look like this:

```ts
slot[modalType === 'non-modal' ? 'required' : 'optional'](action, {
  defaultProps: {
    // bunch of props...
  }
})
```

I agree that the whole `required` flag has been a source of confusion in the whole API, and I'd love to have additional method to make this more specific, but cases like the one of the `DialogTitle` IMO is more confusing and aesthetically worst. So I believe that having the `slot` method will still be required for those cases and we could add those extra "helper methods" as a simplification for all the other use cases